File: workflow/kernel/tools/compile_task_spec.py
#!/usr/bin/env python3
"""Compile an RPI task-spec into a validated `.amanar/workflow.json`.

A task-spec is the human front-end to the workflow contract — GOAL, DONE-WHEN,
SCOPE, VERIFY, BLAST-RADIUS — with the controller's per-check machinery filled by
defaults. The compiled contract is validated with the kernel's own schema and
written canonically. This is an authoring helper, not part of the frozen CLI.

Usage: `python3 .amanar/kernel/tools/compile_task_spec.py SPEC.json [--out PATH]`
"""
from __future__ import annotations

import argparse
import json
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))  # kernel dir -> amanar_workflow

from amanar_workflow.contract import validate  # noqa: E402
from amanar_workflow.errors import ContractError  # noqa: E402

SPEC_FIELDS = {"id", "goal", "scope", "blastRadius", "artifacts", "verify"}
BLAST_FIELDS = {"writes", "liveEffects", "exclusions"}
VERIFY_FIELDS = {"id", "run", "expectedExit", "contains", "timeout", "minTests", "parser", "liveEffect"}


class SpecError(Exception):
    """A task-spec that is malformed before it reaches contract validation."""


def _reject_unknown(value: object, allowed: set[str], label: str) -> dict:
    if not isinstance(value, dict):
        raise SpecError(f"{label} must be an object")
    unknown = sorted(set(value) - allowed)
    if unknown:
        raise SpecError(f"{label} has unknown fields: {', '.join(unknown)}")
    return value


def compile_spec(spec: object) -> dict:
    _reject_unknown(spec, SPEC_FIELDS, "task spec")
    assert isinstance(spec, dict)
    for required in ("id", "goal", "scope", "verify"):
        if required not in spec:
            raise SpecError(f"task spec missing required field: {required}")
    blast = _reject_unknown(spec.get("blastRadius", {}), BLAST_FIELDS, "blastRadius")

    verify = spec["verify"]
    if not isinstance(verify, list) or not verify:
        raise SpecError("verify must be a non-empty array")
    checks = []
    for entry in verify:
        _reject_unknown(entry, VERIFY_FIELDS, "verify entry")
        for required in ("id", "run"):
            if required not in entry:
                raise SpecError(f"verify entry missing required field: {required}")
        checks.append({
            "id": entry["id"],
            "command": entry["run"],
            "expectedExit": entry.get("expectedExit", 0),
            "outputContains": entry.get("contains", []),
            "timeoutSeconds": entry.get("timeout", 120),
            "minTests": entry.get("minTests", 0),
            "testParser": entry.get("parser", "none"),
            "liveEffect": entry.get("liveEffect", False),
        })

    contract = {
        "schemaVersion": "1.0.0",
        "id": spec["id"],
        "objective": spec["goal"],
        "scope": spec["scope"],
        "exclusions": blast.get("exclusions", []),
        "artifacts": spec.get("artifacts", []),
        "authority": {
            "repositoryWrites": blast.get("writes", True),
            "liveEffects": blast.get("liveEffects", False),
        },
        "checks": checks,
    }
    return validate(contract)


def main(argv: list[str] | None = None) -> None:
    parser = argparse.ArgumentParser(prog="compile_task_spec")
    parser.add_argument("spec", help="task-spec JSON file")
    parser.add_argument("--out", default=".amanar/workflow.json", help="output contract path")
    args = parser.parse_args(argv)
    try:
        spec = json.loads(Path(args.spec).read_text(encoding="utf-8"))
    except (OSError, UnicodeError, json.JSONDecodeError) as exc:
        print(f"compile_task_spec error: cannot read spec: {exc}", file=sys.stderr)
        raise SystemExit(1) from exc
    try:
        contract = compile_spec(spec)
    except (SpecError, ContractError) as exc:
        print(f"compile_task_spec error: {exc}", file=sys.stderr)
        raise SystemExit(1) from exc
    out = Path(args.out)
    out.parent.mkdir(parents=True, exist_ok=True)
    out.write_text(json.dumps(contract, indent=2, sort_keys=True) + "\n", encoding="utf-8")
    print(f"wrote {out}")


if __name__ == "__main__":
    main()

File: workflow/kernel/src/tools/compileTaskSpec.ts
/**
 * Compile an RPI task-spec into a validated .amanar/workflow.json.
 * Mirrors tools/compile_task_spec.py exactly.
 */

import fs from 'node:fs';
import path from 'node:path';
import { validate } from '../contract.ts';
import { ContractError } from '../errors.ts';

const SPEC_FIELDS = new Set(['id', 'goal', 'scope', 'blastRadius', 'artifacts', 'verify']);
const BLAST_FIELDS = new Set(['writes', 'liveEffects', 'exclusions']);
const VERIFY_FIELDS = new Set(['id', 'run', 'expectedExit', 'contains', 'timeout', 'minTests', 'parser', 'liveEffect']);

export class SpecError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'SpecError';
  }
}

function rejectUnknown(value: unknown, allowed: Set<string>, label: string): Record<string, unknown> {
  if (typeof value !== 'object' || value === null || Array.isArray(value)) {
    throw new SpecError(`${label} must be an object`);
  }
  const obj = value as Record<string, unknown>;
  const unknown = Object.keys(obj).filter(k => !allowed.has(k)).sort();
  if (unknown.length > 0) {
    throw new SpecError(`${label} has unknown fields: ${unknown.join(', ')}`);
  }
  return obj;
}

export function compileSpec(spec: unknown): Record<string, unknown> {
  const s = rejectUnknown(spec, SPEC_FIELDS, 'task spec');

  for (const required of ['id', 'goal', 'scope', 'verify']) {
    if (!(required in s)) {
      throw new SpecError(`task spec missing required field: ${required}`);
    }
  }

  const blast = rejectUnknown(s['blastRadius'] ?? {}, BLAST_FIELDS, 'blastRadius');

  const verifyRaw = s['verify'];
  if (!Array.isArray(verifyRaw) || verifyRaw.length === 0) {
    throw new SpecError('verify must be a non-empty array');
  }

  const checks: Record<string, unknown>[] = [];
  for (const entry of verifyRaw) {
    const e = rejectUnknown(entry, VERIFY_FIELDS, 'verify entry');
    for (const required of ['id', 'run']) {
      if (!(required in e)) {
        throw new SpecError(`verify entry missing required field: ${required}`);
      }
    }
    checks.push({
      id: e['id'],
      command: e['run'],
      expectedExit: e['expectedExit'] ?? 0,
      outputContains: e['contains'] ?? [],
      timeoutSeconds: e['timeout'] ?? 120,
      minTests: e['minTests'] ?? 0,
      testParser: e['parser'] ?? 'none',
      liveEffect: e['liveEffect'] ?? false,
    });
  }

  const contract = {
    schemaVersion: '1.0.0',
    id: s['id'],
    objective: s['goal'],
    scope: s['scope'],
    exclusions: (blast['exclusions'] ?? []) as unknown[],
    artifacts: (s['artifacts'] ?? []) as unknown[],
    authority: {
      repositoryWrites: blast['writes'] ?? true,
      liveEffects: blast['liveEffects'] ?? false,
    },
    checks,
  };

  return validate(contract);
}

export function main(argv: string[] = process.argv.slice(2)): void {
  // Parse args: spec [--out PATH]
  const args = argv.slice();
  const outIdx = args.indexOf('--out');
  let outPath = '.amanar/workflow.json';
  if (outIdx !== -1) {
    outPath = args[outIdx + 1];
    args.splice(outIdx, 2);
  }
  const specPath = args[0];
  if (!specPath) {
    process.stderr.write('compile_task_spec error: no spec file given\n');
    process.exit(1);
  }

  let spec: unknown;
  try {
    spec = JSON.parse(fs.readFileSync(specPath, 'utf8'));
  } catch (e) {
    process.stderr.write(`compile_task_spec error: cannot read spec: ${e}\n`);
    process.exit(1);
  }

  let contract: Record<string, unknown>;
  try {
    contract = compileSpec(spec);
  } catch (e) {
    if (e instanceof SpecError || e instanceof ContractError) {
      process.stderr.write(`compile_task_spec error: ${e.message}\n`);
      process.exit(1);
    }
    throw e;
  }

  const out = path.resolve(outPath);
  fs.mkdirSync(path.dirname(out), { recursive: true });
  fs.writeFileSync(
    out,
    JSON.stringify(
      contract,
      (_k, v: unknown) => {
        if (v !== null && typeof v === 'object' && !Array.isArray(v)) {
          const sorted: Record<string, unknown> = {};
          for (const k of Object.keys(v as Record<string, unknown>).sort()) {
            sorted[k] = (v as Record<string, unknown>)[k];
          }
          return sorted;
        }
        return v;
      },
      2,
    ) + '\n',
    'utf8',
  );
  console.log(`wrote ${out}`);
}
